promql benchmarks are broken

**What did you do?**
```
cd promql
got test -run=NONE -bench=Benchmark.
```
**What did you expect to see?**

Run each of the benchmarks

**What did you see instead? Under which circumstances?**

```
panic: cannot unpin already unpinned chunk

goroutine 104 [running]:
panic(0x787f00, 0xc420a9e780)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/prometheus/prometheus/storage/local.(*chunkDesc).unpin(0xc421322880, 0xc421032360)
        /home/tristan/go/src/github.com/prometheus/prometheus/storage/local/chunk.go:150 +0x115
github.com/prometheus/prometheus/storage/local.(*memorySeriesIterator).Close(0xc4210fc380)
        /home/tristan/go/src/github.com/prometheus/prometheus/storage/local/series.go:670 +0x59
github.com/prometheus/prometheus/storage/local.(*boundedIterator).Close(0xc4200f2f60)
        /home/tristan/go/src/github.com/prometheus/prometheus/storage/local/storage.go:482 +0x33
github.com/prometheus/prometheus/promql.(*Engine).closeIterators.func1(0x7f02964d4318, 0xc420678640, 0x40ca0f)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/engine.go:534 +0x96
github.com/prometheus/prometheus/promql.inspector.Visit(0x855870, 0x7f02964d4318, 0xc420678640, 0xf6c43a35000003bf, 0x0)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/ast.go:306 +0x3a
github.com/prometheus/prometheus/promql.Walk(0x9dd820, 0x855870, 0x7f02964d4318, 0xc420678640)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/ast.go:255 +0x7b
github.com/prometheus/prometheus/promql.Walk(0x9dd820, 0x855870, 0x9dd7a0, 0xc4200f3840)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/ast.go:275 +0x234
github.com/prometheus/prometheus/promql.Walk(0x9dd820, 0x855870, 0x7f02964d42c0, 0xc420ec6140)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/ast.go:285 +0x74a
github.com/prometheus/prometheus/promql.Inspect(0x7f02964d42c0, 0xc420ec6140, 0x855870)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/ast.go:316 +0x4b
github.com/prometheus/prometheus/promql.(*Engine).closeIterators(0xc4210c0d80, 0xc4210fa780)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/engine.go:538 +0x67
github.com/prometheus/prometheus/promql.(*Engine).execEvalStmt(0xc4210c0d80, 0x7f02964dd1a8, 0xc4201c87e0, 0xc420ed7df8, 0xc4210fa780, 0x9e03a0, 0xc4200f3820, 0x0, 0x0)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/engine.go:398 +0x11af
github.com/prometheus/prometheus/promql.(*Engine).exec(0xc4210c0d80, 0xc420ed7df8, 0x0, 0x0, 0x0, 0x0)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/engine.go:357 +0x392
github.com/prometheus/prometheus/promql.(*query).Exec(0xc420ed7df8, 0x0)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/engine.go:196 +0x38
github.com/prometheus/prometheus/promql.(*Test).exec(0xc420678550, 0x9dbea0, 0xc42000cc80, 0x9dbda0, 0xc420a9e640)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/test.go:466 +0x1fd
github.com/prometheus/prometheus/promql.(*Test).RunAsBenchmark(0xc420678550, 0xc4200f2240, 0x9dbda0, 0xc420a9e640)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/test.go:422 +0xa3
github.com/prometheus/prometheus/promql.(*Benchmark).Run(0xc4200f2240)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/bench.go:42 +0x65
github.com/prometheus/prometheus/promql.BenchmarkHoltWinters4Week5Min(0xc420084280)
        /home/tristan/go/src/github.com/prometheus/prometheus/promql/functions_test.go:29 +0x4e
testing.(*B).runN(0xc420084280, 0x64)
        /usr/local/go/src/testing/benchmark.go:139 +0xaa
testing.(*B).launch(0xc420084280)
        /usr/local/go/src/testing/benchmark.go:277 +0x118
created by testing.(*B).doBench
        /usr/local/go/src/testing/benchmark.go:245 +0x70
exit status 2
FAIL    github.com/prometheus/prometheus/promql 0.060s
```
Running a git bisect points at
```[3bfec97d46676812d6cb7e4225abd2ba6376d4f3] Make the storage interface higher-level.```

**Environment**

* System information:

Linux pinot 4.7.0-rc7-ARCH

* Prometheus version: 

```
prometheus, version 1.0.1 (branch: HEAD, revision: 3bfec97d46676812d6cb7e4225abd2ba6376d4f3)
  build user:       tristan@pinot
  build date:       20160823-18:47:54
  go version:       go1.7
```
Have had the same error on a go1.6 build, and with 1.0 (via the bisect)

* Prometheus configuration file:

n/a , running benchmarks

* Logs:
```
```

--- a/promql/bench.go
+++ b/promql/bench.go
@@ -39,7 +39,9 @@ func (b *Benchmark) Run() {
 	b.b.ReportAllocs()
 	b.b.ResetTimer()
 	for i := 0; i < b.b.N; i++ {
-		b.t.RunAsBenchmark(b)
+		if err := b.t.RunAsBenchmark(b); err != nil {
+			b.b.Error(err)
+		}
 		b.iterCount++
 	}
 }

--- a/promql/engine.go
+++ b/promql/engine.go
@@ -481,40 +481,30 @@ func (ng *Engine) populateIterators(s *EvalStmt) error {
 	Inspect(s.Expr, func(node Node) bool {
 		switch n := node.(type) {
 		case *VectorSelector:
-			var iterators []local.SeriesIterator
-			var err error
 			if s.Start.Equal(s.End) {
-				iterators, err = ng.querier.QueryInstant(
+				n.iterators, queryErr = ng.querier.QueryInstant(
 					s.Start.Add(-n.Offset),
 					StalenessDelta,
 					n.LabelMatchers...,
 				)
 			} else {
-				iterators, err = ng.querier.QueryRange(
+				n.iterators, queryErr = ng.querier.QueryRange(
 					s.Start.Add(-n.Offset-StalenessDelta),
 					s.End.Add(-n.Offset),
 					n.LabelMatchers...,
 				)
 			}
-			if err != nil {
-				queryErr = err
+			if queryErr != nil {
 				return false
 			}
-			for _, it := range iterators {
-				n.iterators = append(n.iterators, it)
-			}
 		case *MatrixSelector:
-			iterators, err := ng.querier.QueryRange(
+			n.iterators, queryErr = ng.querier.QueryRange(
 				s.Start.Add(-n.Offset-n.Range),
 				s.End.Add(-n.Offset),
 				n.LabelMatchers...,
 			)
-			if err != nil {
-				queryErr = err
+			if queryErr != nil {
 				return false
-			}
-			for _, it := range iterators {
-				n.iterators = append(n.iterators, it)
 			}
 		}
 		return true

--- a/promql/functions.go
+++ b/promql/functions.go
@@ -522,7 +522,7 @@ func funcStddevOverTime(ev *evaluator, args Expressions) model.Value {
 		for _, v := range values {
 			sum += v.Value
 			squaredSum += v.Value * v.Value
-			count += 1
+			count++
 		}
 		avg := sum / count
 		return model.SampleValue(math.Sqrt(float64(squaredSum/count - avg*avg)))
@@ -536,7 +536,7 @@ func funcStdvarOverTime(ev *evaluator, args Expressions) model.Value {
 		for _, v := range values {
 			sum += v.Value
 			squaredSum += v.Value * v.Value
-			count += 1
+			count++
 		}
 		avg := sum / count
 		return squaredSum/count - avg*avg

--- a/promql/functions_test.go
+++ b/promql/functions_test.go
@@ -22,7 +22,7 @@ load 5m
     http_requests{path="/foo"}    0+10x8064
 
 eval instant at 4w holt_winters(http_requests[4w], 0.3, 0.3)
-    {path="/foo"} 20160
+    {path="/foo"} 80640
 `
 
 	bench := NewBenchmark(b, input)
@@ -51,7 +51,7 @@ load 1m
     http_requests{path="/foo"}    0+10x1440
 
 eval instant at 1d holt_winters(http_requests[1d], 0.3, 0.3)
-    {path="/foo"} 20160
+    {path="/foo"} 14400
 `
 
 	bench := NewBenchmark(b, input)
@@ -65,7 +65,7 @@ load 1m
     http_requests{path="/foo"}    0+10x1440
 
 eval instant at 1d changes(http_requests[1d])
-    {path="/foo"} 20160
+    {path="/foo"} 1440
 `
 
 	bench := NewBenchmark(b, input)

